src/mpid/ch4/am/rndv/rndv_pre.h

```diff
@@ -0,0 +1,86 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2019 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ */
+#ifndef MPIDI_AM_RNDV_PRE_H_INCLUDED
+#define MPIDI_AM_RNDV_PRE_H_INCLUDED
+
+/* Rendezvous Active Message headers. This is part of MPIDI_am_hdr_t
+ * which starts with the handler_id used by the progress engine in
+ * the receiver to select the right callback. Follows the protocol
+ * header (rndv in this case), represented as union of available
+ * protocol headers; finally, an ending area that used by the netmod
+ * or shmmod to add netmod/shmmod specific header information.
+ *
+ * struct MPIDI_am_hdr_t {
+ *
+ *     int handler_id;
+ *
+ *     MPIDI_am_proto_hdr_t proto_hdr;
+ *
+ *     union {
+ *         MPIDI_NM_am_hdr_t nm_hdr;
+ *         MPIDI_SHM_am_hdr_t shm_hdr;
+ *     } transport;
```

Comment:
probably better to replace this with a padding area to avoid exposing NM/SHM structures externally

Reply:
Yes!

Reply:
The design idea is each protocol will have its own protocol header, which will be opaque to both transport and other protocol. So rather than a generic "am_hdr" or "proto_hdr", you could simply define `rndv_rts_hdr` and `rndv_ack_hdr` etc. I.e. you can define specific headers for each protocol stage.

Reply:
yes proto_hdr is just a union of protocol headers for different phases of the protocol

Reply:
What I was saying is that you don't need the "union". You can directly use individual header for different phase. It is OK since all these headers will be opaque and all equivalent blackbox to the outside.